src/lib.rs

```diff
@@ -3830,6 +3815,47 @@ fn wait_on_child(cmd: &Command, program: &str, child: &mut Child) -> Result<(),
     }
 }
 
+/// Find the destination object path for each file in the input source files,
+/// and store them in the output Object.
+fn objects_from_files(files: &[Arc<Path>], dst: &Path) -> Result<Vec<Object>, Error> {
+    let mut objects = Vec::with_capacity(files.len());
+    for file in files {
+        let obj = if file.has_root() || file.components().any(|x| x == Component::ParentDir) {
```

Comment:
I'm thinking that we need to do some normalisation before running this line, e.g. `././a` => `a`, `a/../b` => `b`

cc @thomcc

Reply:
Probably. Though it's worth noting that `a/../b` does not necessarily == `b`, in particular if `a` is a symlink.

TBH I think we should just remove the check entirely and always take the codepath that does the hashing. It's simpler, fixes some problems where object filename conflcits can cause issue[^1], and sidesteps potential issues we could have with normalisation.

[^1]: for instance, today, if you try to build a lib with something like `cc::Build::new().file("a/object.c").file("b/object.c").compile("lib")`, it will only contain a single file named object.o - one of the two objects will be discarded.

Reply:
I agree, it's probably a good idea to always hash the prefix.

Reply:
Though I still think we can do some normalisation to remove `./` as it is no-op and not affected by symlink or whatever.

Reply:
I removed the condition, and make it so it _always_ does the hashing.

I'm not sure what the normalization is supposed to achieve, and worry that it'll create more problems than it solves. @NobodyXu can you explain why you want the path to be normalised?

Reply:
I don't think it is necessary now that I think about it again, all we care is unique object file name here.